### ISSUE TITLE
Add workspace name to separate applications in window switcher when using allWorkspaces=yes option

### ISSUE THREAD

**Issue by Flrian** [labwc/labwc#1591]
See title, currently it makes it seems like all applications are on one workspace when using the new option. I think the workspace title should be added above the applications on a different workspace.

I've made an image to better illustrate what I mean:
![current](https://github.com/labwc/labwc/assets/4444593/ffd12ffd-6437-40cf-9454-3176906c7452)
My suggestion:
![suggestion](https://github.com/labwc/labwc/assets/4444593/d9012122-54e6-47b4-b39d-911c37846aeb)

**Comment by droc12345** [comment]
You can try this patch, not exactly what you envision, but the same info in a different way and
it allows to set the workspace prefix (default Workspace) when using numbers instead of names.

https://github.com/droc12345/labwc/commit/46f153320ca5b9f40ba065f9b8a94a74cea8e9cf.patch

**Comment by Consolatis** [comment]
Yes, thought about `2.` as well for some time. Technically we could even go fancy and load other implementations dynamically from an external library path. Downside of that is that plugins will break on refactors if we don't define a API for them. And that in turn sounds like something I'd rather not spend time working on / maintaining at the current point.

In regards to the topic itself: we'd basically need to group the windows by workspaces before processing them, both in the osd and in desktop_cycle_view. Once that is done we can easily just check in the osd if the workspace name of the current view to render differs from the last view and then add a heading.

The implementation by @droc12345 can use the existing fields instead so its much easier to implement and users can configure it.

**Comment by johanmalm** [comment]
> Technically we could even go fancy and load other implementations dynamically from an external library path. Downside of that is that plugins (and thus labwc) will break on refactors if we don't define a API for them. And that in turn sounds like something I'd rather not spend time working on / maintaining at the current point.

We think alike. In tree is better. And I reckon the only reason to ever go plugin with in-tree is if RAM becomes a problem. But for a few `render_osd()` implementations, I just can't see that being a factoring even worth considering.

**Comment by droc12345** [comment]
PR #1611

**Comment by droc12345** [comment]
Had dropped old PR and put in a new one.
That's why I blanked out the previous comments as they had pointers to code that didn't exist anymore.

New code just got merged, has the ability to show workspace the window is on,
as well as state of window (min, max, full) and output name (if more than one output)

Consolatis is working on more enhancements to window switch OSD.

**Comment by Consolatis** [comment]
We do already get the data as an array of structs, to be precise an `wl_array()` of `struct view **`.
Thus we technically could already sort and limit its size but the issue here is that the logic in `desktop_cycle_view()` uses the ordering in the `server->views` `wl_list` as well so we actually select the next item shown in the OSD. When we end up modifying the order in the `wl_array()` used in `osd.c` we get out of sync and its basically random what application is selected on `A-Tab`.
